pkg/controller/controller.go

```diff
@@ -125,9 +241,286 @@ func (c *Controller) Start(wg *sync.WaitGroup) error {
+		// the statefulsets for us.
+		sts.SetOwnerReferences([]metav1.OwnerReference{*selfRef})
+		sts, err = c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).Create(sts)
+		if err != nil {
+			c.logger.Error(err.Error())
+			return err
+		}
+
+		// c.recorder.Event(cluster, corev1.EventTypeNormal, "CreatedSts", "create a statefulset")
+		c.logger.Info("created statefulset", zap.String("name", name))
+		return nil
+	}
+
+	// TODO(schallert): propagate whether services were created back up to client
+	if err := c.ensureServices(cluster); err != nil {
+		return err
+	}
+
+	if !cluster.Status.HasInitializedNamespace() {
+		updated, err := c.validateNamespaceWithStatus(cluster)
+		if err != nil {
+			return err
+		}
+
+		// TODO(schallert): Decide for sure what our conventions for re-enqueueing
+		// will be.
+		if updated {
+			return errors.New("re-enqueing cluster due to api update")
+		}
+	}
+
+	if !cluster.Status.HasInitializedPlacement() {
+		updated, err := c.validatePlacementWithStatus(cluster)
+		if err != nil {
+			return err
+		}
+
+		if updated {
+			return errors.New("re-enqueing cluster due to API update")
+		}
+	}
+
+	c.logger.Info("nothing to do",
+		zap.Int("childrensets", len(childrenSets)),
+		zap.Int("zones", len(zones)),
+		zap.Int64("generation", cluster.ObjectMeta.Generation),
+		zap.String("rv", cluster.ObjectMeta.ResourceVersion))
+
+	return nil
+}
+
+// This func is currently read-only, but if we end up modifying statefulsets
+// we'll have to deepcopy.
+func (c *Controller) getChildStatefulSets(cluster *myspec.M3DBCluster) ([]*appsv1.StatefulSet, error) {
+	// TODO(schallert): mark sts w/ a label selector so we can query more
+	// efficiently (already have label helpers in k8sops)
+	statefulSets, err := c.statefulSetLister.StatefulSets(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error listing statefulsets: %v", err))
+		return nil, err
+	}
+
+	childrenSets := make([]*appsv1.StatefulSet, 0)
+	for _, sts := range statefulSets {
+		if metav1.IsControlledBy(sts, cluster) {
+			childrenSets = append(childrenSets, sts.DeepCopy())
+		}
+	}
+
+	return childrenSets, nil
+}
+
+func (c *Controller) handleStatefulSetUpdate(obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("error decoding tombstone, invalid type"))
+			return
+		}
+
+		c.logger.Info("recovered object from tombstone", zap.String("name", object.GetName()))
+	}
+
+	c.logger.Info("processing statefulset", zap.String("name", object.GetName()))
+
+	owner := metav1.GetControllerOf(object)
+	// TODO(schallert): const
+	if owner == nil || owner.Kind != "m3dbcluster" {
```

Comment:
isn't `Kind` named `M3DBCluster`, however to be honest I wasn't very much a fan of that but code `M3` using `M3DB` everywhere. Naming is hard in programming.

Reply:
The more I think about it, I wonder if it just makes sense to rename this CRD to `Cluster`? I feel like M3DB is implied by the group namespace

Reply:
+1 agreed with switching to `Cluster`

Reply:
I think after this PR lands we should make one to 1) switch to cluster and 2) move the api version from `v1` to `v1alpha` or `v1alpha1`